### PR TITLE
Terraui 453 slide panel implementation

### PR DESCRIPTION
**Pull request by tbiethman**
### Summary
This pull request includes the React and SASS implementation of the SlidePanel component.

It should be up to date with the most recent generator, but I will update it again once the review wraps up.

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE
@cerner/terra

### CODE REVIEW

**Comment by bjankord** [comment]
What are your thoughts on splitting this into 2 PRs?
* One for adding the code from initially running the generator in this repo
* One for the implementation of this component. 

This would make it a bit easier for me to focus on the implementation.

**Comment by tbiethman** [comment]
I did initialize the repo with the generator, but the generator is a bit of a moving target at this point. I can merge in the few additions that came after the fact if that makes things easier.

**Comment by bjankord** [comment]
@tbiethman Yeah I'd merge down to master all the files from generator and send out another PR for the slide panel implementation.

File: src/SlidePanel.jsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    rootAttributes.className,
+  ]);
+
+  rootAttributes['data-slide-panel-panel-behavior'] = props.panelBehavior;
+  rootAttributes['data-slide-panel-panel-position'] = props.panelPosition;
+  rootAttributes['data-slide-panel-panel-size'] = props.panelSize;
+
+  return (
+    <div {...rootAttributes}>
+      <div className="terra-SlidePanel-panel" aria-hidden={!props.isOpen ? 'true' : null}>
```
Comment (tbiethman) [inline src/SlidePanel.jsx]: The ternary operation here is to prevent the aria-hidden attribute from being set to "false" when it should be absent entirely.

File: src/SlidePanel.jsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+SlidePanel.propTypes = {
+  mainContent: React.PropTypes.oneOfType([React.PropTypes.object]),
+  panelContent: React.PropTypes.oneOfType([React.PropTypes.object]),
+  panelBehavior: React.PropTypes.oneOf(['push', 'overlay', 'squish']),
+  panelPosition: React.PropTypes.oneOf(['start', 'end']),
+  panelSize: React.PropTypes.oneOf(['small', 'large']),
+  isFullscreen: React.PropTypes.bool,
+  isOpen: React.PropTypes.bool,
+  fill: React.PropTypes.bool,
```
Comment (tbiethman) [inline src/SlidePanel.jsx]: This 'fill' concept that we talked about last week doesn't really mesh with the 'is' semantics for the other boolean states. I'm up for suggestions here.

I got the 'fill' idea from react-bootstrap: https://react-bootstrap.github.io/components.html#panels-tables

File: stories/SlidePanelDemo.jsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    if (bidiValue === 'ltr') {
+      htmlRoot.setAttribute('dir', 'rtl');
+    } else {
+      htmlRoot.setAttribute('dir', 'ltr');
+    }
+  }
+};
+
+class SlidePanelDemo extends React.Component {
```
Comment (tbiethman) [inline stories/SlidePanelDemo.jsx]: This demo component was made to help manage panel states for the storybook examples. Let me know if you think this is going to far; I don't want to misrepresent what we're providing with the SlidePanel component itself, but I also want to provide reasonable usage examples.

File: src/SlidePanel.jsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      <div className="terra-SlidePanel-main">
+        {props.mainContent}
+      </div>
+    </div>
+  );
+};
+
+SlidePanel.propTypes = {
+  mainContent: React.PropTypes.oneOfType([React.PropTypes.object]),
+  panelContent: React.PropTypes.oneOfType([React.PropTypes.object]),
```
Comment (StephenEsser) [inline src/SlidePanel.jsx]: Should these be restricted to a react element or node?

{code}
mainContent: React.PropTypes.element,
{code}

File: src/SlidePanel.jsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import React from 'react';
```
Comment (StephenEsser) [inline src/SlidePanel.jsx]: If you also import PropTypes here you can drop the React dot syntax. Doesn't provide any new functionality, it just shortens syntax.

So instead of 
```
isFullscreen: React.PropTypes.bool,
```
It could be written as
```
isFullscreen: PropTypes.bool,
```

https://github.com/cerner/terra-button/blob/master/src/Button.jsx#L1
https://github.com/cerner/terra-button/blob/master/src/Button.jsx#L1

File: src/SlidePanel.jsx (hunk 4 above)
Comment (tbiethman) [inline src/SlidePanel.jsx]: Yep

File: src/SlidePanel.jsx (hunk 5 above)
Comment (tbiethman) [inline src/SlidePanel.jsx]: I can do that.

File: src/SlidePanel.jsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+SlidePanel.defaultProps = {
+  panelBehavior: 'overlay',
+  panelPosition: 'end',
+  panelSize: 'small',
+};
+
+export default SlidePanel;
+
+export const includePaths = path.join(__dirname, '../src/');
```
Comment (mhemesath) [inline src/SlidePanel.jsx]: whats this for?

Comment (tbiethman) [inline src/SlidePanel.jsx]: Hmm, it's a holdover from some of the early React prototypes. It's unnecessary, I'll remove it.

File: src/SlidePanel.jsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    customProps.className,
+  ]);
+
+  customProps['data-slide-panel-panel-behavior'] = props.panelBehavior;
+  customProps['data-slide-panel-panel-position'] = props.panelPosition;
+  customProps['data-slide-panel-panel-size'] = props.panelSize;
+
+  return (
+    <div {...customProps}>
+      <div className="terra-SlidePanel-panel" aria-hidden={!props.isOpen ? 'true' : null}>
```
Comment (Matt-Butler) [inline src/SlidePanel.jsx]: `aria-hidden={!props.isOpen ? 'true' : 'false'}`?

aria-hidden=null seems wrong

File: lib/SlidePanel.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+SlidePanel.defaultProps = {
+  panelBehavior: 'overlay',
+  panelPosition: 'end',
+  panelSize: 'small'
+};
+
+exports.default = SlidePanel;
+var includePaths = exports.includePaths = _path2.default.join(__dirname, '../src/');
```
Comment (Matt-Butler) [inline lib/SlidePanel.js]: We can remove the includePaths.  It was originally used for applications to import stylesheets.

**Comment by Matt-Butler** [comment]
https://cl.ly/2u2S3N1N1a2J
Is this start example correct?

File: src/SlidePanel.jsx (hunk 7 above)
Comment (tbiethman) [inline src/SlidePanel.jsx]: That would render out to ```aria-hidden="false"```.

Setting the attribute to null keeps the attribute from rendering completely, which makes more sense to me but does lead to this weird null.

File: src/SlidePanel.jsx (hunk 6 above)
Comment (Matt-Butler) [inline src/SlidePanel.jsx]: It can be removed. It was used for importing stylesheets. We don't need it anymore

File: lib/SlidePanel.js (hunk 8 above)
Comment (tbiethman) [inline lib/SlidePanel.js]: It should be removed from the jsx file, I'll make sure this gets rebuilt.

**Comment by tbiethman** [comment]
There's an issue with the bidirectionality not getting defaulted. Try toggling the direction and see if it  looks better. I'm looking into a fix for that.

File: src/slide-panel.scss [hunk 9]
```diff
@@ -0,0 +1,188 @@
+@import 'node_modules/terra-mixins/src/terra-mixins';
+@import './variables';
+@import './mixins';
+
+.terra-SlidePanel {
```
Comment (Matt-Butler) [inline src/slide-panel.scss]: output:
https://gist.github.cerner.com/MB034780/4ec5bf8172d710cbd62387c45a316a08

This file is generating a lot of long css selectors. It seems like we should be able to decrease this a bit.

What are your thoughts? 

@bjankord @mhemesath

File: src/slide-panel.scss [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    &.terra-SlidePanel--is-open[data-slide-panel-panel-position='end'] > .terra-SlidePanel-panel {
+      @include terra-slide-panel-border-start();
+    }
+  }
+}
+
+// Fullscreen
+.terra-SlidePanel--is-fullscreen {
+  > .terra-SlidePanel-panel {
+    width: 100% !important;
```
Comment (Matt-Butler) [inline src/slide-panel.scss]: I would also recommend we don't use !important anywhere unless we have to.

File: src/SlidePanel.jsx [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+const ObjectFilter = (obj, predicate) => {
+  const filtered = Object
+                     .keys(obj)
+                     .filter(key => predicate(key, obj[key]))
+                     .map(key => ({ [key]: obj[key] }));
+  return Object.assign({}, ...filtered);
+};
+
+const CustomProps = (props, propTypes) =>
```
Comment (mjhenkes) [inline src/SlidePanel.jsx]: reminder to replace this with the util method when its created

Comment (tbiethman) [inline src/SlidePanel.jsx]: I'll log an issue once this is merged.

File: src/slide-panel.scss (hunk 9 above)
Comment (Matt-Butler) [inline src/slide-panel.scss]: I would expect classes similar to this:

Component Classes:

.terra-SlidePanel
.terra-SlidePanel-main
.terra-SlidePanel-panel

Modifier Classes:
.terra-SlidePanel.is-open

.terra-SlidePanel-panel--start
.terra-SlidePanel-panel--end
.terra-SlidePanel-panel--small
.terra-SlidePanel-panel--large
.terra-SlidePanel-panel--fullscreen

Each css class should have a single responsibility. Classes should compose into the styles you need.  

Getting rid of the descendent selectors will make this code much smaller too.

File: src/slide-panel.scss (hunk 10 above)
Comment (tbiethman) [inline src/slide-panel.scss]: I hear you. There are multiple selectors of increased complexity that would override this one, and I felt that using important was more efficient than adding ```:not(.terra-SlidePanel--is-fullscreen)``` to the previous selectors or adding additional complex overrides here. 

I'd be happy to try another method if someone is strongly against it.

File: src/slide-panel.scss (hunk 9 above)
Comment (tbiethman) [inline src/slide-panel.scss]: Let me see what I can do with what you're proposing; if I can rewrite it fairly easily, I'm not opposed to doing so. But given our accelerated timetable, I don't want to increase too much entropy in concepts that are already developed.

Comment (tbiethman) [inline src/slide-panel.scss]: Unfortunately, my attempts at adding more modifiers and reducing descendent selectors seemed to just increase complexity, both in the stylesheet and in the React component itself.

One issue is that we would also need modifiers for the .terra-SlidePanel-main component as well, as we style it differently for the various behaviors. On top of that, unless we build unique modifiers for the different permutations (--squishStartSmall, --squishEndSmall, etc.), we're still going to have a long selectors, just in a different way (```.terra-SlidePanel > .terra-SlidePanel-panel.terra-SlidePanel-panel--end.terra-SlidePanel-panel--small {}```)

Please fill me in if I'm missing something obvious here.

File: docs/themeable-variables.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+## Themeable Variables
+
+Variables to `theme` the Terra SlidePanel component.
+
+| Variable Name | Default Value | Usage |
+|---|---|---|
+| `$terra-slide-panel-background-color` | `$terra-white` | Sets the background color of both content and panel areas.|
```
Comment (mhemesath) [inline docs/themeable-variables.md]: i don't feel like this should be done by the slide panel... seems like this could be accomplished with a utility IMO

File: src/_mixins.scss [hunk 13]
```diff
@@ -0,0 +1,63 @@
+@import './variables';
+
+@mixin terra-slide-panel-border-start() {
```
Comment (mhemesath) [inline src/_mixins.scss]: terra should have mixins that handle this already....

File: src/_mixins.scss [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    [dir='rtl'] & {
+      border-right-style: solid;
+    }
+  } @else {
+    border-left-style: solid;
+  }
+}
+
+@mixin terra-slide-panel-border-end() {
```
Comment (mhemesath) [inline src/_mixins.scss]: terra should have mixins to do this already

File: src/_mixins.scss [hunk 15]
```diff
@@ -0,0 +1,63 @@
+@import './variables';
```
Comment (mhemesath) [inline src/_mixins.scss]: really this whole file should be common utities that slidepanel just calls with the property it wants to set

File: src/_variables.scss [hunk 16]
```diff
@@ -0,0 +1,5 @@
+@import 'node_modules/terra-legacy-theme/src/terra-legacy-theme';
+
+$terra-slide-panel-background-color: $terra-white !default;
```
Comment (mhemesath) [inline src/_variables.scss]: slide panels shoudl be transparent IMO... if a background color needs to be set we should allow that to be passed to the panel

File: src/slide-panel.scss [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+@import './variables';
+@import './mixins';
+
+.terra-SlidePanel {
+  overflow: hidden;
+  position: relative;
+
+  > .terra-SlidePanel-main,
+  > .terra-SlidePanel-panel {
+    background-color: $terra-slide-panel-background-color;
```
Comment (mhemesath) [inline src/slide-panel.scss]: I think this should be transparent.... utilities can be used to set this to any color needed. Making it themable doesn't really help if you have mulitple slide panels that need different colors.

File: src/slide-panel.scss [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+    z-index: 0;
+  }
+
+  > .terra-SlidePanel-panel {
+    position: absolute;
+    z-index: 500; // z-index used to set master above detail in visual stacking context.
+  }
+
+  // Prevent z-index issues with nesting
+  .terra-SlidePanel > .terra-SlidePanel-panel {
```
Comment (mhemesath) [inline src/slide-panel.scss]: why is this needed? By definition of the naming convention `.terra-SlidPanel-panel` is a descendant of `.terra-SlidePanel` so this is redundant.

File: src/slide-panel.scss (hunk 10 above)
Comment (mhemesath) [inline src/slide-panel.scss]: what are the other selectors conflicting?

File: src/slide-panel.scss (hunk 18 above)
Comment (tbiethman) [inline src/slide-panel.scss]: It's for nested scenarios; this selector is nested within the `.terra-SlidePanel` selector on line 5.

This selector boils down to:
```
.terra-SlidePanel .terra-SlidePanel > .terra-SlidePanel-panel {...}
```

So, the z-index of a panel of a slide panel within a slide panel is set to 499. 

This is how the current Terra SlidePanel is written and really only protects against one level of nesting; I'm assuming its only used when the panels go fullscreen in smaller viewports.

File: src/slide-panel.scss (hunk 10 above)
Comment (tbiethman) [inline src/slide-panel.scss]: I think I've removed a couple unnecessary selectors that make removing the important here fairly simple. The following is sufficient to override in all scenarios:

```
.terra-SlidePanel--is-fullscreen {
  &[data-slide-panel-panel-size='small'] > .terra-SlidePanel-panel,
  &[data-slide-panel-panel-size='large'] > .terra-SlidePanel-panel {
    width: 100%;
  }
}
```

File: src/slide-panel.scss (hunk 18 above)
Comment (mhemesath) [inline src/slide-panel.scss]: sorry trying to understand.. what problem is this solving. I get its changing nested z-indexing, but why do we need ot do that?

File: src/slide-panel.scss (hunk 10 above)
Comment (mhemesath) [inline src/slide-panel.scss]:
```scss
.terra-SlidePanel--is-fullscreen[data-slide-panel-panel-size] > .terra-SlidePanel-panel {
    width: 100%;
}
```

File: src/_mixins.scss (hunk 15 above)
Comment (tbiethman) [inline src/_mixins.scss]: Doh, I missed the border utilities. I can use those directly. There aren't any current box-shadow utilities, but I can write up a generic one to contribute to terra-mixins.

File: src/slide-panel.scss (hunk 18 above)
Comment (tbiethman) [inline src/slide-panel.scss]: Say you have a SlidePanel inside another SlidePanel. Both SlidePanels are open. On a large device, this should look fine-ish (the overlay behavior could make this weird, but that's a different problem). 

But in small viewports, since both panels are open, both slide panels go fullscreen. We're wanting to ensure that the outermost panel displays on top of any inner ones. Hence the inner panel z-index being bumped down a notch.

I've added a story to the storybook page to demonstrate this situation.

File: src/slide-panel.scss (hunk 17 above)
Comment (tbiethman) [inline src/slide-panel.scss]: I can see -main being transparent, but the -panel element really needs to have a background color, particularly when it is overlaid on top of the main content. 

I'd prefer to default a color like we're doing here (to make the most simple use case completely functional) and force consumers to set background colors on the content they're passing in if they're not happy with the default. Or we could not provide a default at all and make it transparent; it will obvious pretty early for consumers if a background color will be necessary for their use case.

I'd rather not expose additional attributes -main and -panel elements in the API.

**Comment by mhemesath** [comment]
👍

File: src/SlidePanel.jsx [hunk 19]
```diff
@@ -0,0 +1,63 @@
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+
+import '../src/slide-panel.scss';
```
Comment (windse7en) [inline src/SlidePanel.jsx]: Why not simplify this to `./slide-panel.scss`?

Comment (tbiethman) [inline src/SlidePanel.jsx]: We use this roundabout path so that it works with the built file that goes into the /lib directory.
